File: Ashutosh-Singh/2D-Array/Spiral-Matrix-II.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        vector<vector<int>> matrix(n, vector<int>(n, 1));
       
        int count = 0;
        int total = n * n; 
        int value = 1;

        // Indexes for the boundaries of the matrix
        int left = 0; 
        int right = n - 1;
        int top = 0;
        int down = n - 1;

        while(count < total) {
            // Traverse the top row
            for (int i = left; count < total && i <= right; i++) {
                matrix[top][i] = value;
                count++;
                value++;
            }
            top++;

            // Traverse the right column
            for (int i = top; count < total && i <= down; i++) {
                matrix[i][right] = value;
                count++;
                value++;
            }
            right--;

            // Traverse the bottom row
            for (int i = right; count < total && i >= left; i--) {
                matrix[down][i] = value;
                count++;
                value++;
            }
            down--;

            // Traverse the left column
            for (int i = down; count < total && i >= top; i--) {
                matrix[i][left] = value;
                count++;
                value++;
            }
            left++;
        }
        return matrix;
    }
};

int main() {
    Solution solution;
    
    int n;
    
    // Input size for the matrix
    cout << "Enter the size of the matrix (n x n): ";
    cin >> n;
    
    // Get the matrix generated in spiral order
    vector<vector<int>> result = solution.generateMatrix(n);

    // Output the generated matrix
    cout << "Generated matrix in spiral order: " << endl;
    for (const auto& row : result) {
        for (int val : row) {
            cout << val << " ";
        }
        cout << endl;
    }

    return 0;
}

File: Anya_Seth/2D/Spiral Matrix II.java
class Solution {
    public int[][] generateMatrix(int n) {
        int r1 = 0;
        int c1 = 0;
        int r2 = n-1;
        int c2  = n-1;
        int[][] matrix = new int[n][n];
        int add = 1;

        while(r1<=r2 && c1<=c2){
            for(int c = c1 ; c <= c2 ; c++){
                matrix[r1][c] = add++;
            }
            for(int r = r1 + 1 ; r <= r2 ; r++){
                matrix[r][c2] = add++;
            }
            if(r1<r2 && c1<c2){
                for(int c  = c2-1 ; c >= c1 ; c-- ){
                    matrix[r2][c] = add++;
                }
                for(int r  = r2-1 ; r > r1 ; r-- ){
                    matrix[r][c1] = add++;
                }

            } 
            r1++;
            c1++;
            r2--;
            c2--;
        }
        return matrix;
    }
}
